cmake/caches/Travis.cmake

```diff
@@ -34,3 +34,6 @@ set(OMR_THR_CUSTOM_SPIN_OPTIONS ON CACHE BOOL "")
 set(OMR_THR_FORK_SUPPORT ON CACHE BOOL "")
 set(OMR_THR_SPIN_WAKE_CONTROL ON CACHE BOOL "")
 set(OMR_THR_THREE_TIER_LOCKING ON CACHE BOOL "")
+
+set(OMR_USE_NATIVE_ENCODING ON CACHE BOOL "")
+set(OMR_TOOLS_USE_NATIVE_ENCODING ON CACHE BOOL "")
```

Comment:
The old (autotools) defaults for both of these was `OFF`.

Reply:
@dnakamura can you comment on these? This commes from the latest two commits.

Reply:
@dnakamura ping?

Reply:
the default for cmake is still off. `Travis.cmake` specifies options we set for travis/ci builds

Reply:
The native encoding properties still say `ON` despite the apparent agreement above.

Reply:
@fjeremic Are planning to fix this? I think it should say:
```
set(OMR_USE_NATIVE_ENCODING OFF CACHE BOOL "")
set(OMR_TOOLS_USE_NATIVE_ENCODING OFF CACHE BOOL "")
```

Reply:
That will turn off those variables for the CI builds which is not what we want. We want to enable those two flags on the builds but keep them disabled otherwise. IMO I would personally just enable them everywhere as it's the correct thing to do.

Reply:
I understood the meaning of 'USE_NATIVE_ENCODING' to mean 'use EBCDIC on z/OS and ASCII elsewhere', in which case, it's *not* the right thing for the travis build (on z/OS).

Reply:
> it's _not_ the right thing for the travis build (on z/OS).

Can you please explain why you believe that? Why would we not want to use EBCDIC on z/OS?

Reply:
Yeah as @fjeremic and @dnakamura said, these options only affect CI jobs, the [default is off](https://github.com/eclipse/omr/pull/3693/files#diff-370bf43c6501eaf4d5fde675c00cbc5fR51). In the old autotools system, these options enabled in CI through the [example config](https://github.com/eclipse/omr/blob/master/example/glue/configure_includes/configure_zos_390.mk#L36). To me, this looks like the same behaviour as before.

Reply:
The `--enable-native-encoding` configuration option is not the same thing as `USE_NATIVE_ENCODING`, see [1]. We do want `--enable-native-encoding` on z/OS, but we don't want `USE_NATIVE_ENCODING` specified globally. Some makefiles say `USE_NATIVE_ENCODING := 1` but many do not.

[1] https://github.com/eclipse/omr/blob/master/omrmakefiles/rules.zos.mk#L23

Reply:
I'm not sure what the answer here is or what is wanted. As far as the _CMake_ OMR build system is concerned we want native encoding enabled by default as files are being checked out via Git for z/OS (Rocket Software). In such a configuration all files are file tagged as ISO8859-1 and because of `_BPX_AUTOCVT=ON` all disk operations get transparently converted from ISO8859-1 to IBM1047 transparently.

As such I cannot envision a scenario in which we would not want to use the native encoding always, since everything will be in EBCDIC in the z/OS case. The CMake z/OS system will not work unless `OMR_EBCDIC` is defined (it gets defined via `USE_NATIVE_ENCODING` in this case).

Reply:
@keithc-ca Can you explain how setting `OMR_USE_NATIVE_ENCODING` and `OMR_TOOLS_USE_NATIVE_ENCODING` to `ON` affects the CI builds, compared to the autotools builds? Given that the genie builds have passed (several times), I'm trying to udnerstand whether any difference between the autotools and cmake builds actually matter.

Reply:
You'll notice that tools like `hookgen`, `tracegen` and `tracemerge` set `USE_NATIVE_ENCODING := 1` in their makefiles. On the other hand no other makefiles do because the expectation is that text files will be read and written as ASCII (given OMR's heritage as part of J9).

`OMR_TOOLS_USE_NATIVE_ENCODING` seems to be specific to cmake builds and seems to be based on the incorrect assumption that all tools will use the same encoding (`ddrgen` is an exception (a tool not located in the `tools` folder) in that it needs to universally operate in ASCII).

I agree that the scenario that @fjeremic describes is a reasonable one (if we have our OMR blinders on), but I don't think his description of `_BPX_AUTOCVT=ON` is accurate.

I suggest it should be the choice of the consuming project (e.g. openj9) as to which parts of OMR interpret text files as ASCII or EBCDIC. This particular file relates only to the OMR travis build, but the concepts encoded in `OMR_USE_NATIVE_ENCODING` and `OMR_TOOLS_USE_NATIVE_ENCODING ` don't align with the autotools capabilities and freedoms.

Reply:
I'll try to summarize this as best as I can from my current understanding.

The OMR project defines a target called j9a2e [1]. This target defines conversion functions between the ASCII and EBCDIC (default on z/OS) encodings, but that's not all. It also overrides libc APIs by defining header files which will replace standard library calls [2] with wrapper calls in the j9a2e target which will transparently under the hood convert ASCII inputs into EBCDIC before actually calling the real standard library functions (example in [3] for the `fopen` API).

The purpose of this target is so that when OpenJ9 consumes OMR it can link against libj9a2e.so and treat everything internally as ASCII which the files on disk remain EBCDIC. This is presumably so that targets which are built around ASCII (such as the natives in OpenJDK) function correctly. Please correct me if I'm wrong on this one as it is a pure guess on my part.

Now naturally if all internal operations are assuming ASCII and literal strings inside the compilation unit also have to be in ASCII so things are consistent. This is why we specify the `-Wc,"convlit(ISO8859-1)"` compiler flag [4] so that it converts all literal strings in every file into ASCII.

> The old (autotools) defaults for both of these was OFF.

Now the default for `USE_NATIVE_ENCODING` as seen in [4] is 0. This is precisely because we only ever built 8 targets in the autotools build system. These are the 8 targets we currently run in our CI (omralgotest, omrvmtest, omrutiltest, omrsigtest, omrrastest, omrsubscribertest, omrtraceoptiontest). **Note** that targets such as jitbuilder, tril, and comptest are not part of that list. The reason being is because these target _cannot_ work unless `USE_NATIVE_ENCODING = 1` because of the heavy use of C++ streams.

For example in [5] we use `std::cout` to print something to the screen. If `USE_NATIVE_ENCODING = 0` then this object file gets compiled with the `-Wc,"convlit(ISO8859-1)"` flag and links against libj9a2e.so. However C++ streams do not operate through libc and as such the functions we override in [2] will not be executed. Now because of `convlit` that literal string is encoded as ASCII and thus we attempt to print an ASCII string to the screen on z/OS which uses EBCIDC as the native encoding. The result of this on my tests is that the terminal session I execute such a simple program in simply hangs. I'm not yet sure if it's because of my terminal or something deeper went wrong. Regardless the issue is well understood.

---

To me the sane solution for OMR is to set `USE_NATIVE_ENCODING = 1` by default so that we use the native encoding of EBCIDC on z/OS because otherwise we may be restricting consumers of OMR into only using the OMR port library to do any IO, not to mention that such consumers of OMR will be limited to not using C++ IO features.

This is a heavy handed restriction IMO and not one I would advise.

For projects such as OpenJ9 which want to consume OMR and work internally with ASCII are free to do so by simply overriding `USE_NATIVE_ENCODING` to 0. 

Note that as part of this PR there are two defines:
   - `OMR_USE_NATIVE_ENCODING` which determines whether the `omr_base` will use native encoding under the hood (`-DOMR_EBCDIC`) or not (`convlit` + link with libj9a2e.so)
   - `OMR_TOOL_USE_NATIVE_ENCODING` which determines whether the 3 tools (hookgen, tracegen, tracemerge) will use native encoding or not. These tools seem to assume they will be working internally with the native encoding of the platform, so EBCDIC in case of z/OS.

Both of these can be overridden by consumers of OMR.

@keithc-ca @Leonardo2718 @dnakamura @rwy0717 do you agree or disagree? If you disagree please be specific about which part and try to provide an alternative if possible.

[1] https://github.com/eclipse/omr/blob/daa8b14b6a37a1a94aacdde5b0d9578600eec5fe/util/a2e/CMakeLists.txt#L23
[2] https://github.com/eclipse/omr/blob/master/util/a2e/headers/stdio.h#L77-L115
[3] https://github.com/eclipse/omr/blob/daa8b14b6a37a1a94aacdde5b0d9578600eec5fe/util/a2e/atoe.c#L690-L717
[4] https://github.com/eclipse/omr/blob/daa8b14b6a37a1a94aacdde5b0d9578600eec5fe/omrmakefiles/rules.zos.mk#L84-L89
[5] https://github.com/eclipse/omr/blob/daa8b14b6a37a1a94aacdde5b0d9578600eec5fe/jitbuilder/release/cpp/samples/Simple.cpp#L38-L41

Reply:
I agree with @fjeremic.

OMR should not, by default, restrict users from using the tools they know (C++ I/O streams in this case). If a consumer of OMR requires deiferent behaviour, they can override the default.

Reply:
I agree users of the OMR project should be free to select the desired behavior with respect to (text) file encoding.

The fundamental difference (and problem) I see is that under the proposed cmake system, `USE_NATIVE_ENCODING` becomes a *global* choice while today with autotools it is only applied to a few folders within OMR. So OpenJ9, for example, would have to set `USE_NATIVE_ENCODING = OFF` to get the desired behavior of the port library but then tools hookgen, tracegen and tracemerge will work differently. It's possible that OpenJ9 could adapt to that change in behavior, but my main point is just that: it's a change of behavior.

Reply:
According to [4] above it is a global setting, unless I'm misinterpreting the meaning of `GLOBAL_CPPFLAGS` in the autotools build system. Regardless I agree what I'm proposing is a change in behavior which deviates from autotools but enables us to build more OMR targets on z/OS (jitbuilder, tril, comptest, and pretty much the entire rest of OMR) and it standardizes on the notion of using the native encoding everywhere _in OMR_, not mixing bits and pieces to be on or off.

> but then tools hookgen, tracegen and tracemerge will work differently.

This isn't true because `OMR_USE_NATIVE_ENCODING` has no bearing on the use of native encoding for those 3 specific tools. `OMR_TOOLS_USE_NATIVE_ENCODING` is what controls whether we use native encoding there. Example in hookgen in [5].

[5] https://github.com/eclipse/omr/pull/3693/files#diff-5c657e288329114e6f63ea85ebefcaadR32

Reply:
Each folder with a makefile can independently select the appropriate value of `USE_NATIVE_ENCODING` before `rules*.mk` is included.

Reply:
I would argue that having the libraries built as EBCDIC by default is probably the best solution, given that it is also the least surprising. If a consumer wants to override that and use a different character encoding than the OS (and take on all the associated headaches) then they are free to do that. By having it an opt-in means that by doing so, they take responsibility of the repercussions of that decision (eg. not being able to use C++ easily,  having to include `atoe.h` etc.)

Reply:
>Each folder with a makefile can independently select the appropriate value of USE_NATIVE_ENCODING before rules*.mk is included.

True, but its not really a magic bullet. Each folder needs to ensure that every thing it links against was built with the same value of `Each folder with a makefile can independently select the appropriate value of `USE_NATIVE_ENCODING` (unless it doesnt use any api involving strings).  However since it is only ever set by hook/tracegen (which have minimal dependencies) its not that hard to keep track of at the moment